tests/test_single_topology_v3.py

```diff
@@ -0,0 +1,180 @@
+from timemachine.fe.utils import get_romol_conf
+from timemachine.ff import Forcefield
+
+
+def test_phenol():
+    """
+    Test that dummy interactions are setup correctly for a phenol. We want to check that bonds and angles
+    are present when either a single root anchor is provided, or when a root anchor and a neighbor anchor is provided.
+    """
+    mol = Chem.AddHs(Chem.MolFromSmiles("c1ccccc1O"))
+
+    # do not allow H to map to O
+    ff = Forcefield.load_from_file("smirnoff_1_1_0_sc.py")
+
+    # set [O,H] as the dummy group
+    all_idxs, _ = setup_dummy_interactions_from_ff(
+        ff, mol, dummy_group=[6, 12], root_anchor_atom=5, nbor_anchor_atom=None
+    )
+    bond_idxs, angle_idxs, improper_idxs = all_idxs
+
+    assert set(bond_idxs) == set([(5, 6), (6, 12)])
+    assert set(angle_idxs) == set([(5, 6, 12)])
+    assert set(improper_idxs) == set()
+
+    # set [O,H] as the dummy group but allow an extra angle
+    all_idxs, _ = setup_dummy_interactions_from_ff(ff, mol, dummy_group=[6, 12], root_anchor_atom=5, nbor_anchor_atom=0)
+    bond_idxs, angle_idxs, improper_idxs = all_idxs
+
+    assert set(bond_idxs) == set([(5, 6), (6, 12)])
+    assert set(angle_idxs) == set([(5, 6, 12), (0, 5, 6)])
+    assert set(improper_idxs) == set()
+
+    # set [H] as the dummy group
+    all_idxs, _ = setup_dummy_interactions_from_ff(ff, mol, dummy_group=[12], root_anchor_atom=6, nbor_anchor_atom=None)
+    bond_idxs, angle_idxs, improper_idxs = all_idxs
+
+    assert set(bond_idxs) == set([(6, 12)])
+    assert set(angle_idxs) == set()
+    assert set(improper_idxs) == set()
+
+    # set [H] as the dummy group
+    all_idxs, _ = setup_dummy_interactions_from_ff(ff, mol, dummy_group=[12], root_anchor_atom=6, nbor_anchor_atom=5)
+    bond_idxs, angle_idxs, improper_idxs = all_idxs
+
+    assert set(bond_idxs) == set([(6, 12)])
+    assert set(angle_idxs) == set([(5, 6, 12)])
+    assert set(improper_idxs) == set()
+
+    with pytest.raises(single_topology_v3.MissingAngleError):
+        all_idxs, _ = setup_dummy_interactions_from_ff(
+            ff, mol, dummy_group=[12], root_anchor_atom=6, nbor_anchor_atom=4
+        )
+
+
+def test_find_dummy_groups_and_anchors():
+    """
+    Test that we can find the anchors and dummy groups when there's a single core anchor atom. When core bond
+    is broken, we should disable one of the angle atoms.
+    """
+    mol_a = Chem.MolFromSmiles("OCCC")
+    mol_b = Chem.MolFromSmiles("CCCF")
+    core = np.array([[1, 2], [2, 1], [3, 0]])
+
+    dgs, jks = single_topology_v3.find_dummy_groups_and_anchors(mol_a, mol_b, core[:, 0], core[:, 1])
+    assert dgs == [{3}]
+    assert jks == [(2, 1)]
+
+    # angle should swap
+    core = np.array([[1, 2], [2, 0], [3, 1]])
+
+    with pytest.warns(CoreBondChangeWarning):
+        dgs, jks = single_topology_v3.find_dummy_groups_and_anchors(mol_a, mol_b, core[:, 0], core[:, 1])
+        assert dgs == [{3}]
+        assert jks == [(2, None)]
```

Comment:
What does it mean when `k` is `None` here? (this case doesn't seem to be documented in `find_dummy_groups_and_anchors`)

Reply:
k being `None` means that there are no suitable angle terms we can add, this can arise when a core bond is broken.